File: api/routers/preferences.py
import logging
from fastapi import APIRouter, Depends, HTTPException
from api.services.preferences_service import PreferencesService
from api.dependencies import get_current_user_id
from api.schemas.preferences import (
    MoodIconsRequest,
    MoodIconsResponse,
    TimeFormatRequest,
    TimeFormatResponse,
)

logger = logging.getLogger(__name__)


def create_preferences_router(preferences_service: PreferencesService) -> APIRouter:
    router = APIRouter()

    @router.get("/preferences/mood-icons")
    def get_mood_icons(user_id: int = Depends(get_current_user_id)):
        try:
            icons = preferences_service.get_mood_icons(user_id)
            return {"icons": icons}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.put("/preferences/mood-icons", response_model=MoodIconsResponse)
    def save_mood_icons(body: MoodIconsRequest, user_id: int = Depends(get_current_user_id)):
        try:
            validated = {}
            for k, v in body.icons.items():
                try:
                    mood_val = int(k)
                    if 1 <= mood_val <= 5:
                        validated[str(mood_val)] = str(v)
                except (TypeError, ValueError):
                    pass
            preferences_service.save_mood_icons(user_id, validated)
            return {"status": "success", "icons": validated}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.get("/preferences/time-format")
    def get_time_format(user_id: int = Depends(get_current_user_id)):
        try:
            return {"use_24_hour_time": preferences_service.get_use_24_hour_time(user_id)}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.put("/preferences/time-format", response_model=TimeFormatResponse)
    def save_time_format(body: TimeFormatRequest, user_id: int = Depends(get_current_user_id)):
        try:
            preferences_service.save_use_24_hour_time(user_id, body.use_24_hour_time)
            return {"status": "success", "use_24_hour_time": body.use_24_hour_time}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    return router

File: functions/_shared/routes/preferences.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import type { AppEnv } from '../lib/env';
import { createDb } from '../db/client';
import { moodIconsRequestSchema, timeFormatRequestSchema } from '@shared/schemas/preferences';
import {
  getMoodIcons,
  getUse24HourTime,
  saveMoodIcons,
  saveUse24HourTime,
} from '../services/preferences-service';

export const preferencesRoutes = new Hono<AppEnv>();

preferencesRoutes.get('/preferences/mood-icons', async (c) => {
  const userId = c.get('userId');
  const db = createDb(c.env.DB);
  return c.json({ icons: await getMoodIcons(db, userId) });
});

preferencesRoutes.put('/preferences/mood-icons', zValidator('json', moodIconsRequestSchema), async (c) => {
  const userId = c.get('userId');
  const { icons } = c.req.valid('json');
  const db = createDb(c.env.DB);

  const validated: Record<string, string> = {};
  for (const [key, value] of Object.entries(icons)) {
    const moodVal = Number.parseInt(key, 10);
    if (Number.isInteger(moodVal) && moodVal >= 1 && moodVal <= 5) {
      validated[String(moodVal)] = String(value);
    }
  }

  await saveMoodIcons(db, userId, validated);
  return c.json({ status: 'success', icons: validated });
});

preferencesRoutes.get('/preferences/time-format', async (c) => {
  const userId = c.get('userId');
  const db = createDb(c.env.DB);
  return c.json({ use_24_hour_time: await getUse24HourTime(db, userId) });
});

preferencesRoutes.put('/preferences/time-format', zValidator('json', timeFormatRequestSchema), async (c) => {
  const userId = c.get('userId');
  const { use_24_hour_time: use24HourTime } = c.req.valid('json');
  const db = createDb(c.env.DB);

  await saveUse24HourTime(db, userId, use24HourTime);
  return c.json({ status: 'success', use_24_hour_time: use24HourTime });
});
